File: suc/swift/043.swift
import SwiftUI

// MARK: - AppTokens (设计系统)

// 扩展 Color 以支持十六进制初始化，方便颜色定义
extension Color {
    init(hex: UInt, alpha: Double = 1.0) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: alpha
        )
    }
}

// 应用程序的设计令牌，包括颜色、字体、形状、间距和阴影
struct AppTokens {
    struct Colors {
        static let primary = Color(hex: 0xFF00E5FF)
        static let secondary = Color(hex: 0xFF00BFA5)
        static let tertiary = Color(hex: 0xFF69F0AE)
        static let background = Color(hex: 0xFF0A0A0A)
        static let surface = Color(hex: 0xFF121212)
        static let surfaceVariant = Color(hex: 0xFF1E1E1E)
        static let outline = Color(hex: 0xFF2C2C2C)
        static let success = Color(hex: 0xFF00C853)
        static let warning = Color(hex: 0xFFFFD600)
        static let error = Color(hex: 0xFFD50000)
        static let onPrimary = Color(hex: 0xFF0A0A0A)
        static let onSecondary = Color(hex: 0xFF0A0A0A)
        static let onBackground = Color(hex: 0xFFFFFFFF)
        static let onSurface = Color(hex: 0xFFFFFFFF)
    }

    struct TypographyTokens {
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular)
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        static let small: CGFloat = 6
        static let medium: CGFloat = 10
        static let large: CGFloat = 14
    }

    struct Spacing {
        static let sm: CGFloat = 6
        static let md: CGFloat = 10
        static let lg: CGFloat = 14
        static let xl: CGFloat = 20
        static let xxl: CGFloat = 28
    }

    // 阴影规范，用于近似 Android 的 elevation
    struct ShadowSpec {
        let radius: CGFloat // 阴影模糊半径
        let x: CGFloat      // 阴影X轴偏移
        let y: CGFloat      // 阴影Y轴偏移
        let opacity: Double // 阴影不透明度
    }

    struct ElevationMapping {
        // 将 Android 的 elevation (dp) 近似映射到 SwiftUI 的 shadow 属性
        static let level1 = ShadowSpec(radius: 4, x: 0, y: 2, opacity: 0.15)
        static let level2 = ShadowSpec(radius: 8, x: 0, y: 4, opacity: 0.18)
        static let level3 = ShadowSpec(radius: 12, x: 0, y: 6, opacity: 0.2)
    }
}

// MARK: - CartItem 数据模型

struct CartItem: Identifiable {
    let id: Int
    let name: String
    let price: Double
}

// MARK: - RootScreen 视图

struct RootScreen: View {
    let items: [CartItem] = [
        CartItem(id: 1, name: "Neon Keyboard", price: 89.99),
        CartItem(id: 2, name: "Wireless Mouse", price: 49.99),
        CartItem(id: 3, name: "Gaming Headset", price: 109.99)
    ]

    @State private var promo: String = ""

    var total: Double {
        items.reduce(0.0) { $0 + $1.price }
    }

    var body: some View {
        ZStack {
            // 背景渐变
            LinearGradient(
                gradient: Gradient(colors: [AppTokens.Colors.background, AppTokens.Colors.surfaceVariant]),
                startPoint: .top,
                endPoint: .bottom
            )
            .ignoresSafeArea()

            ScrollView {
                VStack(spacing: AppTokens.Spacing.md) {
                    Text("Your Cart")
                        .font(AppTokens.TypographyTokens.display)
                        .foregroundColor(AppTokens.Colors.primary)
                        .frame(maxWidth: .infinity, alignment: .leading)

                    // 购物车商品列表
                    ForEach(items) { item in
                        cartItemCard(item: item)
                    }

                    Spacer().frame(height: AppTokens.Spacing.md)

                    // 促销码输入框
                    HStack {
                        TextField("", text: $promo)
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onSurface)
                            .placeholder(when: promo.isEmpty) {
                                Text("Promo Code")
                                    .font(AppTokens.TypographyTokens.body)
                                    .foregroundColor(AppTokens.Colors.onSurface.opacity(0.5))
                            }
                            .padding(AppTokens.Spacing.md)
                            .background(AppTokens.Colors.surfaceVariant)
                            .cornerRadius(AppTokens.Shapes.medium)
                            .overlay(
                                RoundedRectangle(cornerRadius: AppTokens.Shapes.medium)
                                    .stroke(AppTokens.Colors.outline, lineWidth: 1)
                            )
                    }

                    Spacer().frame(height: AppTokens.Spacing.md)

                    // 总结卡片
                    VStack(spacing: AppTokens.Spacing.sm) {
                        HStack {
                            Text("Items:")
                                .font(AppTokens.TypographyTokens.body)
                                .foregroundColor(AppTokens.Colors.onSurface)
                            Spacer()
                            Text("\(items.count)")
                                .font(AppTokens.TypographyTokens.body)
                                .foregroundColor(AppTokens.Colors.primary)
                        }
                        
                        HStack {
                            Text("Total:")
                                .font(AppTokens.TypographyTokens.title)
                                .foregroundColor(AppTokens.Colors.onSurface)
                            Spacer()
                            Text("$\(String(format: "%.2f", total))")
                                .font(AppTokens.TypographyTokens.title)
                                .foregroundColor(AppTokens.Colors.primary)
                        }
                    }
                    .padding(AppTokens.Spacing.lg)
                    .frame(maxWidth: .infinity)
                    .background(AppTokens.Colors.surfaceVariant)
                    .cornerRadius(AppTokens.Shapes.large)
                    .shadow(
                        color: AppTokens.Colors.background.opacity(AppTokens.ElevationMapping.level1.opacity),
                        radius: AppTokens.ElevationMapping.level1.radius,
                        x: AppTokens.ElevationMapping.level1.x,
                        y: AppTokens.ElevationMapping.level1.y
                    )

                    // 结账按钮
                    Button(action: {}) {
                        Text("Checkout")
                            .font(AppTokens.TypographyTokens.title)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .frame(maxWidth: .infinity)
                            .frame(height: 56)
                            .background(AppTokens.Colors.secondary)
                            .cornerRadius(AppTokens.Shapes.large)
                    }
                    .buttonStyle(PlainButtonStyle())
                }
                .padding(AppTokens.Spacing.lg)
            }
        }
        .statusBar(hidden: true)
    }

    // 购物车商品卡片
    private func cartItemCard(item: CartItem) -> some View {
        HStack(alignment: .center) {
            VStack(alignment: .leading, spacing: 4) {
                Text(item.name)
                    .font(AppTokens.TypographyTokens.title)
                    .foregroundColor(AppTokens.Colors.onSurface)
                Text("$\(String(format: "%.2f", item.price))")
                    .font(AppTokens.TypographyTokens.body)
                    .foregroundColor(AppTokens.Colors.secondary)
            }
            
            Spacer()
            
            Button(action: {}) {
                Text("Remove")
                    .font(AppTokens.TypographyTokens.label)
                    .foregroundColor(AppTokens.Colors.onPrimary)
                    .padding(.horizontal, AppTokens.Spacing.md)
                    .padding(.vertical, AppTokens.Spacing.sm)
                    .background(AppTokens.Colors.primary)
                    .cornerRadius(AppTokens.Shapes.medium)
            }
            .buttonStyle(PlainButtonStyle())
        }
        .padding(AppTokens.Spacing.md)
        .frame(maxWidth: .infinity)
        .background(AppTokens.Colors.surface)
        .cornerRadius(AppTokens.Shapes.large)
        .shadow(
            color: AppTokens.Colors.background.opacity(AppTokens.ElevationMapping.level1.opacity),
            radius: AppTokens.ElevationMapping.level1.radius,
            x: AppTokens.ElevationMapping.level1.x,
            y: AppTokens.ElevationMapping.level1.y
        )
    }
}

// MARK: - 占位符扩展

extension View {
    func placeholder<Content: View>(
        when shouldShow: Bool,
        alignment: Alignment = .leading,
        @ViewBuilder placeholder: () -> Content
    ) -> some View {
        ZStack(alignment: alignment) {
            placeholder().opacity(shouldShow ? 1 : 0)
            self
        }
    }
}

// MARK: - 应用程序入口点

@main
struct ShoppingCartApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
                .preferredColorScheme(.dark)
        }
    }
}

// MARK: - 预览

struct RootScreen_Previews: PreviewProvider {
    static var previews: some View {
        RootScreen()
    }
}

File: suc/kotlin/043.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "057*ShoppingCart*en"
private const val UI_TYPE = "ECommerce"
private const val STYLE_THEME = "Dark Neon"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF00E5FF)
        val secondary = Color(0xFF00BFA5)
        val tertiary = Color(0xFF69F0AE)
        val background = Color(0xFF0A0A0A)
        val surface = Color(0xFF121212)
        val surfaceVariant = Color(0xFF1E1E1E)
        val outline = Color(0xFF2C2C2C)
        val success = Color(0xFF00C853)
        val warning = Color(0xFFFFD600)
        val error = Color(0xFFD50000)
        val onPrimary = Color(0xFF0A0A0A)
        val onSecondary = Color(0xFF0A0A0A)
        val onBackground = Color(0xFFFFFFFF)
        val onSurface = Color(0xFFFFFFFF)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(10.dp)
        val large = RoundedCornerShape(14.dp)
    }
    object Spacing {
        val sm = 6.dp
        val md = 10.dp
        val lg = 14.dp
        val xl = 20.dp
        val xxl = 28.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.15f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.18f)
        val level3 = ShadowSpec(10.dp, 12.dp, 6.dp, 0.2f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

data class CartItem(val id: Int, val name: String, val price: Double)

@Composable
fun RootScreen() {
    val items = remember {
        listOf(
            CartItem(1, "Neon Keyboard", 89.99),
            CartItem(2, "Wireless Mouse", 49.99),
            CartItem(3, "Gaming Headset", 109.99)
        )
    }
    val promo = remember { mutableStateOf("") }
    val total = items.sumOf { it.price }
    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(
                    Brush.verticalGradient(
                        listOf(AppTokens.Colors.background, AppTokens.Colors.surfaceVariant)
                    )
                )
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
        ) {
            Text("Your Cart", style = AppTokens.TypographyTokens.display, color = AppTokens.Colors.primary)
            items.forEach {
                Card(
                    shape = AppTokens.Shapes.large,
                    colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                    elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level1.elevation),
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(AppTokens.Spacing.md),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Column {
                            Text(it.name, style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.onSurface)
                            Text("$${String.format("%.2f", it.price)}", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.secondary)
                        }
                        Button(
                            onClick = {},
                            colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary),
                            shape = AppTokens.Shapes.medium
                        ) {
                            Text("Remove", color = AppTokens.Colors.onPrimary, style = AppTokens.TypographyTokens.label)
                        }
                    }
                }
            }
            Spacer(Modifier.height(AppTokens.Spacing.md))
            TextField(
                value = promo.value,
                onValueChange = { promo.value = it },
                modifier = Modifier.fillMaxWidth(),
                placeholder = { Text("Promo Code", color = AppTokens.Colors.onSurface.copy(alpha = 0.5f)) },
                colors = TextFieldDefaults.colors(
                    focusedContainerColor = AppTokens.Colors.surface,
                    unfocusedContainerColor = AppTokens.Colors.surfaceVariant,
                    focusedIndicatorColor = AppTokens.Colors.primary,
                    unfocusedIndicatorColor = AppTokens.Colors.outline
                ),
                shape = AppTokens.Shapes.medium
            )
            Spacer(Modifier.height(AppTokens.Spacing.md))
            Card(
                shape = AppTokens.Shapes.large,
                colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surfaceVariant),
                elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level1.elevation),
                modifier = Modifier.fillMaxWidth()
            ) {
                Column(
                    modifier = Modifier.padding(AppTokens.Spacing.lg),
                    verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)
                ) {
                    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                        Text("Items:", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.onSurface)
                        Text("${items.size}", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.primary)
                    }
                    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                        Text("Total:", style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.onSurface)
                        Text("$${String.format("%.2f", total)}", style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.primary)
                    }
                }
            }
            Button(
                onClick = {},
                modifier = Modifier.fillMaxWidth().height(56.dp),
                colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.secondary),
                shape = AppTokens.Shapes.large
            ) {
                Text("Checkout", color = AppTokens.Colors.onSecondary, style = AppTokens.TypographyTokens.title)
            }
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFF0A0A0A)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}
